contrib/dynamic_embedding/src/tde/details/cacheline_id_transformer.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+#include <memory>
+#include <optional>
+#include "nlohmann/json.hpp"
+#include "tcb/span.hpp"
+#include "tde/details/move_only_function.h"
+#include "tde/details/naive_id_transformer.h"
+
+namespace tde::details {
+
+template <typename LXURecord>
+struct CachelineIDTransformerValue {
+  int64_t global_id_not_;
+  uint32_t cache_id_;
+  LXURecord lxu_record_;
+};
+
+template <typename LXURecord>
+class CachelineIDTransformerIterator {
+  using Val = CachelineIDTransformerValue<LXURecord>;
+
+ public:
+  CachelineIDTransformerIterator(Val* begin, Val* end)
+      : begin_(begin), end_(end) {}
+
+  std::optional<TransformerRecord<LXURecord>> operator()() {
+    for (; begin_ != end_;) {
+      auto& record = *begin_++;
+      if (record.global_id_not_ >= 0) {
+        continue;
+      }
+      TransformerRecord<LXURecord> result{};
+      result.global_id_ = -record.global_id_not_;
+      result.cache_id_ = record.cache_id_;
+      result.lxu_record_ = record.lxu_record_;
+      return result;
+    }
+    return std::nullopt;
+  }
+
+ private:
+  Val* begin_;
+  Val* end_;
+};
+
+/**
+ * CachelineIDTransformer
+ *
+ * Transform GlobalID to CacheID by naive flat hash map
+ * @tparam LXURecord The extension type used for eviction strategy.
+ * @tparam Bitmap The bitmap class to record the free cache ids.
+ */
+template <
+    typename LXURecord,
+    int64_t NumCacheline = 8,
+    int64_t CachelineSize = 64,
+    typename BitMap = Bitmap<uint32_t>,
+    typename Hash = std::hash<int64_t>>
+class CachelineIDTransformer {
+ public:
+  static_assert(NumCacheline > 0, "NumCacheline should be positive.");
+  static_assert(CachelineSize > 0, "CachelineSize should be positive.");
+  using Self = CachelineIDTransformer<
+      LXURecord,
+      NumCacheline,
+      CachelineSize,
+      BitMap,
+      Hash>;
+  static constexpr std::string_view type_ = "cacheline";
+
+  explicit CachelineIDTransformer(int64_t num_embedding, int64_t capacity = 0);
+
+  CachelineIDTransformer(const Self&) = delete;
+  CachelineIDTransformer(Self&&) noexcept = default;
+
+  static Self Create(int64_t num_embedding, const nlohmann::json& json) {
```

Comment:
nit: In pytorch we use lowercase methods. Lets make it the same that it will look naturally pytorchy.